### PR TITLE
Feature/add dwh singleton

### PR DESCRIPTION
**Pull request by DawidMakar-acaisoft**
The changeset consists of:

- adding Python 3.12 support, add a new GitHub check
- adding a new class `DwhSingleton` - ensures the Data Warehouse class can be instantiated only once
- adding a test case that confirms the above behavior

### CODE REVIEW

File: src/pyprediktorutilities/dwh.py [hunk 1]
```diff
@@ -302,3 +304,11 @@ def __disconnect(self) -> None:
     def __commit(self) -> None:
         """Commits any changes to the database."""
         self.connection.commit()
+
+
+class DwhSingleton(Dwh, metaclass=singleton_class.SingletonMeta):
```
[the remark this answers is not in the document]
Comment (pavel-tashev) [reply src/pyprediktorutilities/dwh.py:309]: @DawidMakar-acaisoft please check the comment above.

[the remark this answers is not in the document]
Comment (DawidMakar-acaisoft) [reply src/pyprediktorutilities/dwh.py:309]: I wasn't sure where we use the `Dwh` class in other applications, so I didn't want to make a breaking change. If you are the only person who uses the utility at this point, I understand that the change works for you well. Thus, I can apply the change just the way you wrote.

File: README.md [hunk 2]
```diff
@@ -141,7 +141,7 @@ dwh = Dwh(SQL_SERVER, SQL_DATABASE, SQL_USER, SQL_PASSWORD, driver_index)
 ## Example usage
 
 ```
-from pyprediktorutilities.dwh import Dwh
+from pyprediktorutilities.dwh.dwh import Dwh
 
 dwh = Dwh("localhost", "mydatabase", "myusername", "mypassword")
 results = dwh.fetch("SELECT * FROM mytable")
```
Comment (pavel-tashev) [inline README.md:147]: In the REAME file also add into about DwhSingleton with example code how to use it. Also, clarify what is the difference between Dwh and DwhSingleton and when to use them.

File: tests/test_dwh.py [hunk 3]
```diff
@@ -1,54 +1,24 @@
...
-    raise pyodbc.DataError("Error code", "Error message")
-
-
-def mock_pyodbc_connection_throws_error_tolerant_to_attempts(connection_string):
-    raise pyodbc.DatabaseError("Error code", "Error message")
-
-
-def grs():
-    """Generate a random string."""
-    return "".join(random.choices(string.ascii_uppercase + string.digits, k=10))
```
Comment (pavel-tashev) [inline tests/test_dwh.py:3]: Inside tests folder create a new folder `dwh` and inside put test_dwh.py and test_dwh_singleton.py. The folder structure of the tests folder must reflect the folder structure of what we test.

**Pull request merged**